dev/generate_requirements.py

```diff
@@ -0,0 +1,112 @@
+    parser = argparse.ArgumentParser(
+        description=(
+            "Generate a pip-compatible requirements.txt file from an MLflow"
+            " requirements.yaml specification"
+        )
+    )
+    parser.add_argument(
+        "--requirements-yaml-location",
+        required=True,
+        help="Local file path of the requirements.yaml specification.",
+    )
+    parser.add_argument(
+        "--requirements-txt-output-location",
+        required=True,
+        help="Local output path for the generated requirements.txt file.",
+    )
+
+    return parser.parse_args(args)
+
+
+def validate_requirements_yaml(requirements_yaml):
+    assert isinstance(
+        requirements_yaml, dict
+    ), "requirements.yaml contents must be a YAML dictionary"
+    for package_entry in requirements_yaml.values():
+        assert isinstance(package_entry, dict), (
+            "Entry in requirements.yaml does not have required dictionary"
+            f" structure: {package_entry}"
+        )
+        pip_release = package_entry.get("pip_release")
+        assert pip_release is not None and isinstance(pip_release, str), (
+            "Entry in requirements.yaml does not define a valid 'pip_release'"
+            f" string value: {package_entry}"
+        )
+        max_major_version = package_entry.get("max_major_version")
+        assert max_major_version is not None and isinstance(max_major_version, int), (
+            "Entry in requirements.yaml does not specify a valid 'max_major_version'"
+            f" integer value: {package_entry}"
+        )
+        if "minimum" in package_entry:
+            assert isinstance(package_entry["minimum"], str), (
+                "Entry in requirements.yaml contains an invalid 'minimum' version"
+                f" string specification: {package_entry}"
+            )
+        if "unsupported" in package_entry:
+            assert isinstance(package_entry["unsupported"], list) and all(
+                [
+                    isinstance(unsupported_entry, str)
+                    for unsupported_entry in package_entry["unsupported"]
+                ]
+            ), (
+                "Entry in requirements.yaml contains an invalid 'unsupported' versions"
+                " specification. Unsupported versions should be specified as lists of strings:"
+                f" {package_entry}"
+            )
+        if "markers" in package_entry:
+            assert isinstance(package_entry["markers"], str), (
+                "Entry in requirements.yaml contains invalid 'markers' string"
+                f" specification: {package_entry}"
+            )
+
+
+def generate_requirements_txt_content(requirements_yaml):
+    requirement_strs = []
+    for package_entry in requirements_yaml.values():
+        pip_release = package_entry["pip_release"]
+        version_specs = []
+
+        max_major_version = package_entry["max_major_version"]
+        version_specs += [f"<{max_major_version + 1}.*"]
```

Comment:
I also tried `f"<= {max_major_version}.*", which does not seem to work (it forces installation of `< {max_major_version}`). From https://peps.python.org/pep-0440/#final-releases the `+ 1` approach seems okay:

```
The release segment consists of one or more non-negative integer values, separated by dots:

N(.N)*

Final releases within a project MUST be numbered in a consistently increasing fashion, otherwise automated tools will not be able to upgrade them correctly.
```

Reply:
I also investigated but it looks like this is the only way to pin a package to `max_major_version` or older.